cppcache/integration/test/CachingProxyTest.cpp

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <list>
+#include <thread>
+
+#include <gtest/gtest.h>
+
+#include <geode/Cache.hpp>
+#include <geode/CacheFactory.hpp>
+#include <geode/RegionFactory.hpp>
+#include <geode/RegionShortcut.hpp>
+#include <geode/Serializable.hpp>
+
+#include "framework/Cluster.h"
+
+using apache::geode::client::Cache;
+using apache::geode::client::Cacheable;
+using apache::geode::client::CacheableString;
+using apache::geode::client::CacheFactory;
+using apache::geode::client::Region;
+using apache::geode::client::RegionShortcut;
+using apache::geode::client::Serializable;
+
+namespace CachingProxy {
+
+class CachingProxyTest : public ::testing::Test {
```

Comment:
You're going to want something like this:

```
protected:
static std::unique_ptr<Cluster> cluster_;
static std::unique_ptr<Cache> cache_;
static std::shared_ptr<Region> region_;

public:
static void SetUpTestSuite() {
  cluster_ = make_unique<Cluster>(Name{"CachingProxyTest"}, LocatorCount{1}, ServerCount{1});

  cluster_->start();

  cluster.getGfsh()
         .create()
         .region()
         .withName("region")
         .withType("PARTITION")
         .execute();

  cache_ = make_unique<Cache>(cluster_->createCache());

  cache_.getPoolManager()
         .createFactory()
         .addLocator("localhost", cluster_.getLocatorPort())
         .create("pool");

  region_ = cache_->createRegionFactory(CACHING_PROXY)
                     .setPoolName("pool")
                     .create("region");
  }

  static void TearDownTestSuite() {
    region_.reset();
    cache_.reset();
    cluster_.reset();
  }
```

Now your cluster is running and configured, your cache is configured, and your region is configured. All will be shared between every test scenario. It will be setup before any scenario is run, it is shutdown after all the scenarios are completed.

And don't forget with statics, you need to define their initial state outside the class declaration. Since they're all smart pointers, this is easy:

```
}; // class CachingProxyTest

std::unique_ptr<Cluster> CachingProxyTest::cluster_{};
std::unique_ptr<Cache> CachingProxyTest::cache_{};
std::shared_ptr<Region> CachingProxyTest::region_{};
```

You can even do this with your test key and test value, and I would recommend you include the test region and test pool names. Don't rely on duplicating literals.

You then only need to be sure to insert the initial value in the constructor, and clear the region in the destructor in preparation for the next test. You'll also be able to get rid of `SetUp` and `TearDown`, which (isn't always the case, but) would be redundant to the ctor and dtor.

Reply:
This is not a pattern you should ever use for testing.

 `Cache`, which is the root of all the components you are integration testing here, should NEVER have shared state between tests. It is too easy to end up writing tests that have indeterminate outcomes based on previous test. Did the test pass because the previous test put it a state it will always pass? Will the order of my tests change the behavior of each test?

Sharing the external system, `Cluster` in this case, is arguably ok to share within a test fixture. I lean very heavily in not for the same reasons above. The previous test framework did this and many tests were inadvertently left in strange states because of it.

Reply:
I have no problem avoiding this pattern for the Cache. However, I think it's worth considering this pattern for the server, which is the expensive part. Also, we're not testing the server here.

Reply:
For moving the `Cluster` initialization to the fixture level to have any advantage you would need also modify the test discovery and execution model in CMake such that those that had a shared `Cluster` were executed serially in the same processes space. Currently all tests are executed in their own process space so any shared effort in the fixture setup is effectively test level setup.